src/main/java/fi/aalto/cs/apluscourses/intellij/services/PluginSettings.java

```diff
@@ -58,26 +58,58 @@ public String getName() {
   private final PropertiesComponent applicationPropertiesManager = PropertiesComponent
       .getInstance();
 
+  private static class ProjectKey {
+    @NotNull
+    public final Project project;
+
+    public ProjectKey(@Nullable Project project) {
+      if (project == null) {
+        this.project = ProjectManager.getInstance().getDefaultProject();
+      } else {
+        this.project = project;
+      }
+    }
+
+    @Override
+    public boolean equals(Object other) {
+      if (other == null || !(other instanceof ProjectKey)) {
+        return false;
+      }
+      return getPath().equals(((ProjectKey) other).getPath());
+    }
+
+    @Override
+    public int hashCode() {
+      return getPath().hashCode();
+    }
+
+    @NotNull
+    private String getPath() {
+      return project.isDefault() ? "" : project.getBasePath();
+    }
+  }
+
```

Comment:
I like this approach a lot. I'm only uncertain if it's fine to a) keep references to possibly disposed `Project` objects (as hash map keys), b) call `isDefault()` and `getBasePath()` methods on them (hash map calls `equals` on it's stored keys). So would it be safer to store just the string returned by `getBasePath`, not the `Project` object itself, what do you think? But it's very good design that the implementation is encapsulated inside a class nonetheless.

Reply:
There is a `projectClosed` listener which removes the entries from the maps, but I agree that it makes more sense to just store the base path (assuming that we don't access the project from the key, which I'll check).

Reply:
Fixed :+1: